pkg/connector/role_cache.go

```diff
@@ -2,106 +2,89 @@ package connector
 
 import (
 	"context"
+	"fmt"
 	"strconv"
-	"sync"
 
+	"github.com/conductorone/baton-sdk/pkg/session"
+	"github.com/conductorone/baton-sdk/pkg/types/sessions"
 	"github.com/conductorone/baton-workato/pkg/connector/client"
 	"github.com/grpc-ecosystem/go-grpc-middleware/logging/zap/ctxzap"
 	"go.uber.org/zap"
 )
 
-type roleCache struct {
-	client       *client.WorkatoClient
-	folderToRole map[int][]*client.Role
-	roles        map[string]*client.Role
-
-	initialized bool
-	mu          sync.Mutex
-}
-
-func newRoleCache(workatoClient *client.WorkatoClient) *roleCache {
-	return &roleCache{
-		client:       workatoClient,
-		folderToRole: make(map[int][]*client.Role),
-		roles:        make(map[string]*client.Role),
-	}
-}
+const (
+	rolesCachePrefix       = "roles"
+	folderRolesCachePrefix = "folder_roles"
+)
 
-func (p *roleCache) init(ctx context.Context) error {
-	p.mu.Lock()
-	defer p.mu.Unlock()
+func getRoleByFolder(ctx context.Context, sessionStorage sessions.SessionStore, folderID string) []*client.Role {
+	l := ctxzap.Extract(ctx)
 
-	if p.initialized {
+	folderRoles, found, err := session.GetJSON[[]*client.Role](ctx, sessionStorage, folderID, sessions.WithPrefix(folderRolesCachePrefix))
+	if err != nil {
+		l.Error("failed to get folder roles from session storage", zap.Error(err))
 		return nil
 	}
 
-	if err := p.buildCache(ctx); err != nil {
-		return err
+	if !found {
+		return nil
 	}
 
-	p.initialized = true
-	return nil
+	return folderRoles
 }
 
-func (p *roleCache) buildCache(ctx context.Context) error {
+func getRoleById(ctx context.Context, sessionStorage sessions.SessionStore, roleID string) *client.Role {
 	l := ctxzap.Extract(ctx)
 
-	l.Info("Building cache for Roles")
+	role, found, err := session.GetJSON[*client.Role](ctx, sessionStorage, roleID, sessions.WithPrefix(rolesCachePrefix))
+	if err != nil {
+		l.Error("failed to get role by id from session storage", zap.Error(err))
+		return nil
+	}
 
-	p.folderToRole = make(map[int][]*client.Role)
-	p.roles = make(map[string]*client.Role)
+	if !found {
+		return nil
+	}
 
-	token := ""
+	return role
+}
 
-	for {
-		roles, nextToken, err := p.client.GetRoles(ctx, token)
-		if err != nil {
-			return err
-		}
+func setRolesCache(ctx context.Context, sessionStorage sessions.SessionStore, roles []client.Role) error {
+	err := session.SetManyJSON(ctx, sessionStorage, parseJSONRolesCache(roles), sessions.WithPrefix("roles"))
+	if err != nil {
+		return fmt.Errorf("failed to set roles cache in session storage: %w", err)
+	}
 
-		for _, role := range roles {
-			copyRole := role
-			for _, folderID := range role.FolderIDs {
-				p.folderToRole[folderID] = append(p.folderToRole[folderID], &copyRole)
+	for _, role := range roles {
+		for _, folderID := range role.FolderIDs {
+			folderIDStr := strconv.Itoa(folderID)
+			folderRoles, found, err := session.GetJSON[[]*client.Role](ctx, sessionStorage, folderIDStr, sessions.WithPrefix("folder_roles"))
```

Comment:
this should be GetJSONMany we should avoid calling session.GetJSON in a loop, it is a network request every time. Recommend just getting everything if you think that makes sense.